### PR TITLE
Finalization Registry proposal

### PR DESCRIPTION
**Pull request by guillep**

### CODE REVIEW

**Comment by Ducasse** [comment]
Thanks for such a description. 
May be we should part of it inside the class comments of the finalizationRegistry.

**Comment by guillep** [comment]
There is already one

https://github.com/pharo-project/pharo/pull/12042/files#diff-97824dad991141332c87f2f47bec03b82c171ca5a11cef03992913963535686f

> 

src/System-Finalization/FinalizationRegistry.class.st

**Comment by Ducasse** [comment]
Nice. I love comments. I always loved comments when I started to code in VW. It was so great to be able to learn.

**Comment by seandenigris** [comment]
Exciting! I know this has been a todo for many years and took a lot of effort - thank you :)

**Review by noha — APPROVED** [review]

**Review by MarcusDenker — APPROVED** [review]

**Review by estebanlm — APPROVED** [review]

**Review by tesonep — APPROVED** [review]

**Review by demarey — APPROVED** [review]

**Comment by estebanlm** [comment]
this phep is approved. 

your id will be 0003, please 
1) rename your file, 
2) add the id to your description 
3) add status: accepted  to your description

... and it will be merged :)

**Comment by privat** [comment]
Nice proposal.

Some questions/remarks:

- what happens when a finalisable object is revived and later recollected. Will the finalisation mechanic called again? (I expect not, but it should be in the spec), if no, can it be readded again to the registry? with a different finalizer? (I expect yes to both)
- is there some constraints on the multiplicity on the finalizable-finaliser relationship? One finalizer for many finalisables? One finalisable for many finalizers? Many to many?
- can a same pair be registered more than once? (I expect that superfluous additions are noops)
- can a pair be removed? (E.g. the finalization stuff was already done (eg a explicit close was performed) thus no more need for the vm/image to track the object)
- what if the finalizer is collected but not the finalisable?
- does it make sence to have multiple registries?
- what are the guarantees that objects are eventually finalized and/or with time, guarantees (I expect that no guarantees for anything, and that an explicit call to the gc may still help)

**Comment by guillep** [comment]
> Nice proposal.

Thanks, I'll enhance the proposal (and implementation) based on this

> Some questions/remarks:
> 
>     * what happens when a finalisable object is revived and later recollected. Will the finalisation mechanic called again? (I expect not, but it should be in the spec),

Only if another finalizer is subscribe for the finalizable.
The subscription that produced the "resurrection" will be discarded and not be called again.

> if no, can it be readded again to the registry? with a different finalizer? (I expect yes to both)
Yes.
Even we could have two finalizers for the same object at the same time.
In that case, we could also think about the order of evaluation of finalizers.

I propose that the order of evaluation remains (explicitly) undefined for this iteration, unless somebody has some strong requirement + cycles to help make it real :).
This would mean that users should not depend on the order.

>     * is there some constraints on the multiplicity on the finalizable-finaliser relationship? One finalizer for many finalisables? One finalisable for many finalizers? Many to many?

My idea was to not give any constraint.
But it's true that under no constraints, user can make mistakes.
If one finalizable has many finalizer, order becomes important.
If many finalizables share the same finalizer, the finalizer should be somehow idempotent, or at least not crash?

> 
>     * can a same pair be registered more than once? (I expect that superfluous additions are noops)

So far my thoughts were that each registration is evaluated on collection.
When the same pair is registered twice, we could make the finalizer to be executed twice.
To me that's the more explicit behaviour.

> 
>     * can a pair be removed? (E.g. the finalization stuff was already done (eg a explicit close was performed) thus no more need for the vm/image to track the object)

I did not consider removal for now, I think it's an addition that could be done later.

> 
>     * what if the finalizer is collected but not the finalisable?

Cannot be, the finalizer is strongly held by the registry, so it cannot be collected :)

> 
>     * does it make sence to have multiple registries?

In terms of maintenance of registries, one is probably more than enough.
I'd leave the door open to have many registries for debugging and testing purposes.
It's not clear to me if we could benefit by splitting registries in smaller registries...

> 
>     * what are the guarantees that objects are eventually finalized and/or with time, guarantees (I expect that no guarantees for anything, and that an explicit call to the gc may still help)

No temporal guarantees :).
I'd say that the implementation should guarantee that, if the object becomes unreachable, the finalizer will be eventually evaluated.
(or at least, users should write their programs under that assumption)

Thoughts?

File: phep-proposal.md [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+## Finalization Overview and Protocol
+
+Finalization is managed at the user level by a finalization registry.
+The finalization registry contains (finalizable, finalizer) pairs.
+The registry has two main methods: `#add:` and `#add:finalizer:`.
+Finalizers need to implement `#finalize` with their corresponding finalization
+
+- `FinalizationRegistry>>#add:` adds an object as both finalizable and finalizer. This means the object is its own finalizer.
+- `FinalizationRegistry>>#add:finalizer:` adds an object as finalizable with another object as finalizer.
+- `[User Object]>>#finalize` invokes the finalization mechanism of the given finalizer. The default `#finalize` method in `Object` is defined to do nothing.
```
Comment (privat) [inline phep-proposal.md:47]: To be consistent with the lack of constraints on multiplicity, I would have a parameter here that is the finalizable object given with `add:finalizer:` (or self if plain `add:` is used). ie `UserFinaliserClass>>#finalize: aUserFinalizableObject`

File: phep-proposal.md [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+"Create a finalizable object implementing #finalize"
+MyFinalizer << Object
+	package: 'MyPackage'.
+MyFinalizer >> #finalize [
+	'Gone!' traceCr.
+]
+
+"Add it to a registry"
+objectOfInterest := Object new.
+registry := FinalizationRegistry new.
```
Comment (privat) [inline phep-proposal.md:63]: use `default` here instead of `new`?

**Comment by privat** [comment]
Thanks for your answers.

Some additional thoughts:

* maybe add the removal thing in the spec. If you expect people to use it to manage their resources, it could be nice to have a full-fledged spec.
* The whole behavior about multiple registration / multiple finalization should be more precise. I understand it is mostly corner case behavior, but as a basic programmer I like when rules are explicit.

Maybe what is missing is a user guide "How to have nice finalization programming". Do you have some existing usage, like for file management (I think that leaks of open system file handlers it is the most obvious use case of a finalization mechanism) that could be given as a good example? Or more generally, a guide for any FFI based handlers?

Especially, since there are no guarantees, the document should be more explicit that the whole finalization proposal is about having a less bad behavior in case of object leaks. And that proper resource management should be correctly taken into account when designing and programming software and API. Basically, this automatic finalization mechanism is meant for when something drops the ball. It's especially important in Pharo because anything lives in the same image, including aborted threads and failed tests and that we expect that the GC helps us to automatically release most of the unneeded resources.

I'ts also why I think that the 'removing' feature is important. If I design something that need resource management, I would like to have the control of when having the seatbelt on. E.g. I allocate something and wrap it and before giving it to the client, so `FinalizationRegistry>>#add:`. And when to have it off. E.g. the client says to me they are done with the resource, so I can do the cleanup, do `Something>>#remove:` on the finalization, and ideally invalidate the resource to fail-fast in case of use-after-free issue, because I still do trust the client :p

Also, it could be nice to have some kind of 'related work' information (it's classic in other languages *EP). For instance, how do they deal with this finalization issue in other languages and in other Smalltalk dialects.

**Pull request merged**